securesystemslib/interface.py

```diff
@@ -607,46 +619,36 @@ def import_ed25519_privatekey_from_file(filepath, password=None):
   securesystemslib.formats.PATH_SCHEMA.check_match(filepath)
 
-  # If the caller does not provide a password argument, prompt for one.
-  # Password confirmation disabled here, which should ideally happen only
-  # when creating encrypted key files (i.e., improve usability).
-  if password is None: # pragma: no cover
+  if password and prompt:
+    raise ValueError("Passing 'password' and 'prompt' True is not allowed.")
+
+  # If 'password' was passed check format and that it is not empty.
+  if password is not None:
+    securesystemslib.formats.PASSWORD_SCHEMA.check_match(password)
 
+    # TODO: PASSWORD_SCHEMA should be securesystemslib.schema.AnyString(min=1)
+    if not len(password):
+      raise ValueError('Password must be 1 or more characters')
+
+  elif prompt:
+    # Password confirmation disabled here, which should ideally happen only
+    # when creating encrypted key files (i.e., improve usability).
     # It is safe to specify the full path of 'filepath' in the prompt and not
     # worry about leaking sensitive information about the key's location.
     # However, care should be taken when including the full path in exceptions
     # and log files.
-    password = get_password('Enter a password for the encrypted Ed25519'
-        ' key (' + Fore.RED + filepath + Fore.RESET + '): ',
-        confirm=False)
-
-  # Does 'password' have the correct format?
-  securesystemslib.formats.PASSWORD_SCHEMA.check_match(password)
-
-  # Store the encrypted contents of 'filepath' prior to calling the decryption
-  # routine.
-  encrypted_key = None
-
-  with open(filepath, 'rb') as file_object:
-    encrypted_key = file_object.read()
-
-  # Decrypt the loaded key file, calling the 'cryptography' library to generate
-  # the derived encryption key from 'password'.  Raise
-  # 'securesystemslib.exceptions.CryptoError' if the decryption fails.
-  key_object = securesystemslib.keys.decrypt_key(encrypted_key.decode('utf-8'),
-      password)
-
-  # Raise an exception if an unexpected key type is imported.
-  if key_object['keytype'] != 'ed25519':
-    message = 'Invalid key type loaded: ' + repr(key_object['keytype'])
-    raise securesystemslib.exceptions.FormatError(message)
-
-  # Add "keyid_hash_algorithms" so that equal ed25519 keys with
-  # different keyids can be associated using supported keyid_hash_algorithms.
-  key_object['keyid_hash_algorithms'] = \
-      securesystemslib.settings.HASH_ALGORITHMS
+    # NOTE: A user who gets prompted for a password, can only signal that the
+    # key is not encrypted by entering no password in the prompt, as opposed
+    # to a programmer who can call the function with or without a 'password'.
+    # Hence, we treat an empty password here, as if no 'password' was passed.
+    password = get_password('Enter a password for an encrypted RSA'
+        ' file \'' + Fore.RED + filepath + Fore.RESET + '\': ',
+        confirm=False) or None
```

Comment:
This `or None` is confusing.  We have a guideline (although specific for outputting text) that hints against this sort of short circuit evaluation.

https://github.com/secure-systems-lab/code-style-guidelines#never-use-short-circuit-evaluation-to-produce-output

Reply:
`password` should be None if the user doesn't enter a password, correct?

Reply:
I'm afraid this code is already there for RSA, too...

https://github.com/secure-systems-lab/securesystemslib/blob/1c4161e0a8b249e42f4f8d02ffef49a45a2e4006/securesystemslib/interface.py#L328-L330

Reply:
Hmm, true.  I see that @awwad  had a similar comment against it [here](https://github.com/secure-systems-lab/securesystemslib/pull/124#discussion_r174238179).  I recommend we avoid copying it.

Reply:
Got it. What do you recommend?

Reply:
I'm not sure why @lukpueh added the `or None`.  Is it even needed?  If the user doesn't enter anything, the empty string is considered None, right?

@awwad, what do you think?

Reply:
Just my 0.02 BTC, but on a separate note, there's a lot of code repeated between both of these functions for RSA and Ed25519 (and, in the future, for ECDSA) that really should be abstracted out and shared...

Reply:
get_password() would return an empty string, and later code tests for `None`, so yes, the `or None` (or some other, more intuitive construction) is necessary.

Reply:
Necessary?  I'm confused, isn't the empty string and None similar in Boolean equivalence?  Is this really necessary, or is it needed for clarity (which I don't find clear at all)?

Reply:
IIRC, there are specific `is None` tests in later code, but it's been a few weeks, so perhaps I'm misremembering.

Reply:
@trishankatdatadog You can add a line to set password to None if len(password) == 0.  That should make it clear.